### PR TITLE
Add watcher args

### PR DESCRIPTION
**Pull request by isvoboda**
This change provides the arguments to Watcher object in watch function.
For example, when I would like to define ignored directories instead of hard-coding them.

~~~python
class ImageWatcher(AllWatcher):
    re_img = re.compile(r".*(jpe?g|png|bmp)$", re.IGNORECASE)

    def __init__(self, root_path, ignore=None):
        self.ignore = ignore or []
        super(ImageWatcher, self).__init__(root_path)

    def should_watch_file(self, entry):
        return self.re_img.match(entry.path)

    def should_watch_dir(self, entry):
        return entry.name not in self.ignore


for changes in watch(".", watcher_cls=ImageWatcher, watcher_args=["ignore"]):
    print(changes)
~~~

### CODE REVIEW

**Comment by samuelcolvin** [comment]
I sorry for the slow reply. I'm currently away traveling and will look soon.

File: watchgod/watcher.py [hunk 1]
```diff
@@ -78,3 +78,16 @@ def should_watch_file(self, entry):
 class PythonWatcher(DefaultDirWatcher):
     def should_watch_file(self, entry):
         return entry.name.endswith(('.py', '.pyx', '.pyd'))
+
+
+class RegExpWatcher(AllWatcher):
+    def __init__(self, root_path, re_files=None, re_dirs=None):
+        self.re_files = re.compile(re_files) if re_files is not None else re_files
+        self.re_dirs = re.compile(re_dirs) if re_files is not None else re_dirs
+        super(RegExpWatcher, self).__init__(root_path)
```
Comment (samuelcolvin) [inline watchgod/watcher.py]: just 

```py
super().__init__(root_path)
```

would be cleaner.

File: watchgod/main.py [hunk 2]
```diff
@@ -52,12 +52,13 @@ class awatch:
...
     __slots__ = (
-        '_loop', '_path', '_watcher_cls', '_debounce', '_min_sleep', '_stop_event', '_normal_sleep', '_w', 'lock',
-        '_executor'
+        '_loop', '_path', '_watcher_cls', '_watcher_args', '_debounce', '_min_sleep', '_stop_event', '_normal_sleep',
+        '_w', 'lock', '_executor'
     )
 
     def __init__(self, path: Union[Path, str], *,
                  watcher_cls: Type[AllWatcher]=DefaultWatcher,
+                 watcher_args: Optional[Sequence]=None,
```
Comment (samuelcolvin) [inline watchgod/main.py]: Would be clearer I think to change this to `watcher_kwargs: Optional[Dict[str, Any]]`

Comment (samuelcolvin) [reply watchgod/main.py]: that also avoid the confusion that this is really "extra_watcher_args".

File: tests/test_watch.py [hunk 3]
```diff
@@ -91,6 +91,29 @@ def test_python(tmpdir):
     assert watcher.check() == {(Change.modified, str(tmpdir.join('foo/spam.py')))}
 
 
+def test_regexp(tmpdir):
+    mktree(tmpdir, tree)
+
+    re_files = r"^.*(\.txt|\.js)$"
```
Comment (samuelcolvin) [inline tests/test_watch.py]: single quotes on strings please, unless double quotes are required.

File: watchgod/main.py (hunk 2 above)
Comment (isvoboda) [reply watchgod/main.py]: How about having both `watcher_kwargs` and `extra_watcher_args`?

Comment (samuelcolvin) [reply watchgod/main.py]: I think over kill, `watcher_kwargs` should suffice in all situations.

**Comment by samuelcolvin** [comment]
great, thanks.

**Pull request merged**